### DISCUSSION TITLE
How do I start a span with no parent when another span is active?

### DISCUSSION THREAD

**Discussion by alevenberg** [open-telemetry/opentelemetry-cpp#2410] [Q&A]
I want to create a span with no parent.

Currently, when I call `StartSpan()`, the parent defaults to the current active span.

I want to do something like
```
  opentelemetry::trace::StartSpanOptions options;
  options.parent = null;
  auto span = tracer->StartSpan("span", {}, {} options);
```

But the parent needs to be `SpanContext` or `context::Context`. By default it uses, `SpanContext::GetInvalid()`, which will create a child of the active span. So, I'm not sure what I else I could use.

Is there a way to start a span with no parent?

Thanks, Anna

**Comment by lalitb** [comment]
For the root span, the parent_span_id field would contain the span_id of the current span. There is no way to make it null/empty/missing.

**Reply by alevenberg** [reply]
Ah that makes sense! Is there a way to start a span such that the parent_span_id gets set to the span_id of the current span?

**Reply by lalitb** [reply]
Sorry, I think I misunderstood your question earlier. You want to create a root span, while a given span is already active - this is not possible, and not sure if this is the right use case either. All the span/log operations are bound to the current active span as a parent.

```cpp

**Reply by alevenberg** [reply]
The use case is for asynchronous batching.

There are N spans that are initiating an async rpc, but the async rpc can occur during one of those spans. Instead of creating the parent/child relationship, we want to represent the relationship with links.

I think you can do this in other languages. I'm not sure about the specifics, but I think it would look like creating a [new span context](https://pkg.go.dev/go.opentelemetry.io/otel/trace#NewSpanContext) and then [starting a new span](https://pkg.go.dev/go.opentelemetry.io/otel/trace#ContextWithSpanContext) with that context.

**Reply by lalitb** [reply]
Yes, on further thought this seems to be a valid requirement. When one can create a span that is a child to a non-active span, it should be possible to create a span with no parent. Would you like to create an issue for this? This can be an ABI-breaking change, so could be targeted for ABI v2.

**Reply by lalitb** [reply]
OK, have created an issue here - https://github.com/open-telemetry/opentelemetry-cpp/issues/2413. Feel free to further discuss there.

**Reply by alevenberg** [reply]
Thanks for making the issue!
